pkg/daemon/daemon.go

```diff
@@ -1 +1,107 @@
 package daemon
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/Mellanox/nic-configuration-operator/api/v1alpha1"
+	"github.com/Mellanox/nic-configuration-operator/pkg/host"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type NicConfigurationDaemon struct {
+	client      client.Client
+	nodeName    string
+	hostManager host.HostManager
+	hostUtils   host.HostUtils
+}
+
+func (d *NicConfigurationDaemon) Run() error {
+	log.Log.Info("Started daemon")
+
+	hostUtils := host.NewHostUtils()
+
+	unchroot, err := hostUtils.Chroot("/host")
+	if err != nil {
+		log.Log.Error(err, "Failed to chroot")
+		return nil
+	}
+	defer unchroot()
+
+	hostManager := host.NewHostManager(d.nodeName, hostUtils)
+
+	devices := hostManager.DiscoverNicDevices()
+
+	for _, deviceStatus := range devices {
+		log.Log.Info("Device found", "deviceStatus", deviceStatus)
+
+		name := deviceStatus.Type + "-" + strings.ToLower(deviceStatus.SerialNumber)
+		namespace := "default"
+
+		device, err := d.createDeviceCRIfNotExists(name, namespace)
+		if err != nil {
+			log.Log.Error(err, "Failed to obtain NicDevice obj")
+			continue
+		}
+
+		err = d.UpdateDeviceStatus(device, deviceStatus)
+		if err != nil {
+			log.Log.Error(err, "Failed to update NicDevice status")
+			continue
+		}
+	}
+
+	time.Sleep(time.Second * 1000)
+	return nil
+}
+
+func (d *NicConfigurationDaemon) createDeviceCRIfNotExists(name string, namespace string) (*v1alpha1.NicDevice, error) {
+	existingDevice := &v1alpha1.NicDevice{}
+	err := d.client.Get(context.TODO(), client.ObjectKey{
```

Comment:
we do want some valid context here : e.g a context that is cancelled if signal recieved.